include/CL/cl_ext_pocl.h

```diff
@@ -60,42 +60,73 @@ typedef CL_API_ENTRY cl_int
 
 #endif
 
-/* cl_pocl_pinned_buffers (experimental stage) */
+/* cl_ext_buffer_device_address (experimental stage) */
 
-#ifndef cl_pocl_pinned_buffers
-#define cl_pocl_pinned_buffers 1
-#define CL_POCL_PINNED_BUFFERS_EXTENSION_NAME "cl_exp_pinned_buffers"
+#ifndef cl_ext_buffer_device_address
+#define cl_ext_buffer_device_address 1
 
-/* TODO: We need also platform/runtime extension due to the new buffer
-   creation flags. */
+/* We need both a platform and device extension due to the new buffer
+   creation flag. */
 
-/* clCreateBuffer(): A new cl_mem_flag CL_MEM_PINNED:
+/* clCreateBuffer(): A new cl_mem_flag CL_MEM_DEVICE_ADDRESS:
 
-   This flag specifies that the buffer must be persistently allocated
-   in the device's physical memory for its lifetime. That is, the buffer's
-   device address will remain the same and the space is reserved until
-   the buffer is freed. The device-specific buffer content updates are
-   still performed by implicit or explicit buffer migrations performed by
-   the runtime or the client code. If any of the devices in the context
-   does not support pinning, an error (TO DEFINE) is returned.
+   This flag specifies that the buffer must have a single fixed address
+   for its lifetime, which should be unique at least across the devices
+   of the context, but not necessarily with the host (virtual memory) as
+   with SVM or USM.
+
+   The flag might imply that the buffer will be "pinned" permanently to
+   a device's memory, but might not be necessarily so, as long as the address
+   range of the buffer remains constant.
+
+   The address is guaranteed to remain the same until the buffer is freed and
+   the address can be queried via clGetMemObjectInfo().
+
+   The device-specific buffer content updates must be still performed by
+   implicit or explicit buffer migrations performed by the runtime or the
+   client code. If any of the devices in the context does not support
+   such allocations, an error (CL_INVALID_VALUE) is returned.
+
+   ***Update to the clCreateSubBuffer()***
+
+   Sub-buffers created from buffers allocated with the flag are guaranteed
+   to be synchronized fully as a single entity, regardless of the offset value
+   passed. That is, when the parent buffer has been allocated with
+   CL_MEM_DEVICE_ADDRESS, setting the size of the sub-buffer to 0 does not
+   return CL_INVALID_BUFFER_SIZE, but denotes that the sub-buffer's
+   parent buffer's contents might be accessed anywhere by the kernel.
+   This is to ensure pointer arithmetics can be done freely within the parent
+   allocation on the kernel side although the kernel would get a pointer to
+   a middle passed in as an argument.
```

Comment:
I don't think this is such a good idea, because that's really isn't a sub-buffer anymore.

Maybe adding `clSetKernelArgDeviceAddressPointer` (similar to `clSetKernelArgSVMPointer`) would allow for a better programming model instead here, as you only need to keep track of your pointers and the runtime can map that to the allocations internally as this tracking needs to happen for `CL_KERNEL_EXEC_INFO_DEVICE_PTRS` already anyway.

This somewhat makes it a new SVM variant, but that doesn't have to be a bad thing if the API gets duplicated. Maybe it can be merged, but for that I'd wait for WG review and what people there decide.

Reply:
Yes, creating a new `clSetKernelArgDeviceAddressPointer` is the cleaner alternative, but since it's much more intrusive spec-vise (adds a new function to the ICD dispatch table, etc.), I'd go with the sub-buffer approach in the draft version and perhaps introduce the new function in the final version of the extension. Is the sub-buffer approach a lot of work to implement in RustiCL/Mesa?

Reply:
Also as long as it stays an extension you'd go through `clGetExtensionFunctionAddress` anyway, so as long as it's not considered to become a core spec it doesn't really matter all that much. But maybe you could also allow `clSetKernelArgSVMPointer` to be used, because it's really fundamentally the same. Could also be renamed to a more generic name via an union in a later CL spec version.

I doubt the amount of work changes that much, it just feels too hacky for my test and makes device-address-buffer sub-buffers objects you'd need to special case in a couple of places. Where if you design it like `clSetKernelArgSVMPointer` runtimes could just share a bunch of code, because it's essentially the same for the runtime, just instead of tracking pointer -> SVM alloc, you'd track pointer -> cl_mem internally or some base + sub classes.

It also allows application to set arbitrary pointers as arguments which would remove the need to constantly allocate new sub-buffers as you can just point to a random pointer directly.

Though maybe there is also merit in creating sub-buffers and using some of the memory Copy and Map APIs on them.

So maybe it's fine to simply bring it up in WG discussions and see how others feel about this.

Reply:
Yes, you are right. I didn't remember that the `clGetExtensionFunctionAddress` indirection helps here. I'll add the new setarg API, makes it much cleaner. It think has to be a new API as we must be able to differentiate between dev ptrs and vm ptrs in case the device supports both?

My current plan with this extension is to prototype it as an 'ext' extension in PoCL (CPU obviously works already, I'm working on L0 implementation right now) and (if I understood your intentions right) in RustiCL, with chipStar as the first example client app. This way we support a lot of CUDA/HIP apps already using only the simple extension. Then if it seems more widely useful (e.g. a lot of OpenMP offload apps should work with it alone) with positive feedback, bring it to OpenCL WG for potential 'khr' promotion discussions.

Reply:
> It think has to be a new API as we must be able to differentiate between dev ptrs and vm ptrs in case the device supports both?

Depends on how runtimes are willing to figure this out. Those allocations can't overlap so you can easily tell them apart, but of course there is runtime overhead you wouldn't have when having two APIs.

> My current plan with this extension is to prototype it as an 'ext' extension in PoCL (CPU obviously works already, I'm working on L0 implementation right now) and (if I understood your intentions right) in RustiCL, with chipStar as the first example client app. This way we support a lot of CUDA/HIP apps already using only the simple extension. Then if it seems more widely useful (e.g. a lot of OpenMP offload apps should work with it alone) with positive feedback, bring it to OpenCL WG for potential 'khr' promotion discussions.

sounds like a plan

Reply:
I sent a PR of the extension text for review feedback: https://github.com/KhronosGroup/OpenCL-Docs/pull/1159 @karolherbst @linehill  pls provide feedback.